cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,1209 @@
+      return visitor.Visit(static_cast<BooleanBuilder*>(builder));
+    case Type::UINT8:
+      return visitor.Visit(static_cast<UInt8Builder*>(builder));
+    case Type::INT8:
+      return visitor.Visit(static_cast<Int8Builder*>(builder));
+    case Type::UINT16:
+      return visitor.Visit(static_cast<UInt16Builder*>(builder));
+    case Type::INT16:
+      return visitor.Visit(static_cast<Int16Builder*>(builder));
+    case Type::UINT32:
+      return visitor.Visit(static_cast<UInt32Builder*>(builder));
+    case Type::INT32:
+      return visitor.Visit(static_cast<Int32Builder*>(builder));
+    case Type::UINT64:
+      return visitor.Visit(static_cast<UInt64Builder*>(builder));
+    case Type::INT64:
+      return visitor.Visit(static_cast<Int64Builder*>(builder));
+    case Type::HALF_FLOAT:
+      return visitor.Visit(static_cast<HalfFloatBuilder*>(builder));
+    case Type::FLOAT:
+      return visitor.Visit(static_cast<FloatBuilder*>(builder));
+    case Type::DOUBLE:
+      return visitor.Visit(static_cast<DoubleBuilder*>(builder));
+    case Type::STRING:
+      return visitor.Visit(static_cast<StringBuilder*>(builder));
+    case Type::BINARY:
+      return visitor.Visit(static_cast<BinaryBuilder*>(builder));
+    case Type::FIXED_SIZE_BINARY:
+      return visitor.Visit(static_cast<FixedSizeBinaryBuilder*>(builder));
+    case Type::DATE32:
+      return visitor.Visit(static_cast<Date32Builder*>(builder));
+    case Type::DATE64:
+      return visitor.Visit(static_cast<Date64Builder*>(builder));
+    case Type::TIMESTAMP:
+      return visitor.Visit(static_cast<TimestampBuilder*>(builder));
+    case Type::TIME32:
+      return visitor.Visit(static_cast<Time32Builder*>(builder));
+    case Type::TIME64:
+      return visitor.Visit(static_cast<Time64Builder*>(builder));
+    case Type::INTERVAL:
+      return Status::NotImplemented("No IntervalBuilder");
+    case Type::DECIMAL:
+      return visitor.Visit(static_cast<Decimal128Builder*>(builder));
+    case Type::LIST:
+      return visitor.Visit(static_cast<ListBuilder*>(builder));
+    case Type::STRUCT:
+      return visitor.Visit(static_cast<StructBuilder*>(builder));
+    case Type::UNION:
+      return Status::NotImplemented("No UnionBuilder");
+    case Type::DICTIONARY:
+      return Status::NotImplemented("No DictionaryBuilder");
+    case Type::MAP:
+      return Status::NotImplemented("No MapBuilder");
+    default:
+      return Status::NotImplemented("Unknown Builder");
+  }
+}
+
+class TypedHandler
+    : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, TypedHandler> {
+ public:
+  TypedHandler(StructBuilder* root_builder) : builder_(root_builder) {}
+
+  bool Null() {
+    if (Skipping()) return true;
+    // TODO check whether we're currently nullable
+    status_ = VisitBuilder(builder_, AppendNullVisitor{});
+    return status_.ok();
+  }
+
+  bool Bool(bool value) {
+    if (Skipping()) return true;
+    if (builder_->type()->id() == Type::BOOL) {
```

Comment:
Or simply have dedicated handler classes (with the approach of restarting the current block if the previous type failed converting).

Reply:
I'll move the type_id to a member variable in the conversion stage. Since I'll be moving that out of SAX handling, the the dedicated handler class won't need to worry about conversion.